### PR TITLE
Implementing Rest API

### PR DESCRIPTION
**Pull request by mkmeral**
This pull requests implements Rest API endpoints and authentication filter.

### CODE REVIEW

**Comment by mkmeral** [comment]
Pull request https://github.com/ibissource/iaf/pull/1461 contains related configurations for api.

File: pom.xml [hunk 1]
```diff
@@ -105,6 +112,36 @@
...
+      <version>${cxf.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.cxf</groupId>
+      <artifactId>cxf-rt-transports-http</artifactId>
+      <version>${cxf.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
```
Comment (jacodg) [inline pom.xml:137]: Shouldn't this dependency have scope provided?

File: src/main/java/nl/nn/testtool/storage/file/Reader.java [hunk 2]
```diff
@@ -133,7 +132,7 @@ protected List getMetadata(int numberOfRecords, List metadataNames,
 			metadataReadOnly = metadataCacheReadOnly;
 		}
 		List result = new ArrayList();
-		for (int i = 0; i < metadataReadOnly.size() &&  (numberOfRecords == -1 || i < numberOfRecords); i++) {
+		for (int i = 0; i < metadataReadOnly.size() &&  (numberOfRecords == -1 || result.size() < numberOfRecords); i++) {
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/storage/file/Reader.java:135]: When result = 5 and requested numberOfRecords = 10 nothing will be returned?
Maybe it's better to do a separate PR which describes the bug you are trying to fix?

File: src/main/java/nl/nn/testtool/util/Export.java [hunk 3]
```diff
@@ -105,7 +115,13 @@ private static ExportResult export(Storage storage, Report report,
 				Set duplicateCheck = new HashSet();
 				Iterator iterator = storageIds.iterator();
 				while (iterator.hasNext()) {
+					try {
 					report = storage.getReport((Integer)iterator.next());
+						if (report == null)
+							continue;
+					} catch (Exception e) {
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/util/Export.java:122]: Should we just ignore any error? ExportResult has setErrorMessage(), maybe we should use that method?

File: src/main/java/nl/nn/testtool/web/FrontendServlet.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+public class FrontendServlet extends HttpServlet {
+
+	private static final long serialVersionUID = 123L;
+	private Map<String, MediaType> mediaTypeMap;
+	private final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+	@Override
+	public void init() throws ServletException {
+		super.init();
+		System.err.println("INITIALIZING FRONTEND SERVER!!!!!!");
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/web/FrontendServlet.java:29]: I think we should remove this before merging to master?

File: src/main/java/nl/nn/testtool/web/FrontendServlet.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		for (String key : properties.stringPropertyNames()) {
+			String value = properties.getProperty(key);
+			mediaTypeMap.put(key, MediaType.valueOf(value));
+		}
+	}
+
+	@Override
+	protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+		String path = req.getPathInfo();
+		System.err.println("DO GET WITH PATH : " + path);
```
Comment (jacodg) [inline src/main/java/nl/nn/testtool/web/FrontendServlet.java:55]: Should also be removed

Comment (celmoussaoui) [reply src/main/java/nl/nn/testtool/web/FrontendServlet.java]: This has been removed

File: src/main/java/nl/nn/testtool/web/FrontendServlet.java (hunk 4 above)
Comment (celmoussaoui) [reply src/main/java/nl/nn/testtool/web/FrontendServlet.java]: This has been removed

File: src/main/java/nl/nn/testtool/util/Export.java (hunk 3 above)
Comment (celmoussaoui) [reply src/main/java/nl/nn/testtool/util/Export.java:122]: I have added the setErrorMessage to the exception

File: pom.xml (hunk 1 above)
Comment (celmoussaoui) [reply pom.xml:137]: Added the provided scope

**Comment by mkmeral** [comment]
@jacodg  I think this PR is good to merge codewise. Also, we also need to merge [this PR](https://github.com/ibissource/iaf/pull/1461).  

However, one thing I have realized is API is fully exposed right now. We need to configure AuthenticationInterceptor, and for that, I'll need your help.

**Review by jacodg — APPROVED** [review]
